File: Python/Builder/builder_2.py
"""
Design pattern: Builder
"""


class Person:
    """
    Person class
    """

    def __init__(self) -> None:
        """
        Constructor
        """
        # Address
        self.street_address: str
        self.postcode: str
        self.city: str
        # Employment
        self.company_name: str
        self.position: str
        self.annual_income: int

    def __str__(self) -> str:
        """
        String representation
        """
        return (
            f"Address: {self.street_address}, {self.postcode}, {self.city}.\n"
            + f"Employed at {self.company_name} as a {self.position} earning {self.annual_income}"
        )


class PersonBuilder:
    """
    Person builder
    """

    def __init__(self, person: Person = Person()) -> None:
        """
        Constructor
        """
        self.person = person

    @property
    def works(self) -> "PersonJobBuilder":
        """
        Works
        """
        return PersonJobBuilder(self.person)

    @property
    def lives(self) -> "PersonAddressBuilder":
        """
        Works
        """
        return PersonAddressBuilder(self.person)

    def build(self) -> Person:
        """
        Build method
        """
        return self.person


class PersonJobBuilder(PersonBuilder):
    """
    Person job builder
    """

    def __init__(self, person: Person) -> None:
        """
        Constructor
        """
        super().__init__(person)

    def at(self, company_name: str) -> "PersonJobBuilder":
        """
        Add company
        """
        self.person.company_name = company_name
        return self

    def as_a(self, position: str) -> "PersonJobBuilder":
        """
        Add position
        """
        self.person.position = position
        return self

    def earning(self, annual_income: int) -> "PersonJobBuilder":
        """
        Add annual income
        """
        self.person.annual_income = annual_income
        return self


class PersonAddressBuilder(PersonBuilder):
    """
    Person addres builder
    """

    def __init__(self, person: Person) -> None:
        """
        Constructor
        """
        super().__init__(person)

    def at(self, street_address: str) -> "PersonAddressBuilder":
        """
        Add street address
        """
        self.person.street_address = street_address
        return self

    def with_postcode(self, postcode: str) -> "PersonAddressBuilder":
        """
        Add postcode
        """
        self.person.postcode = postcode
        return self

    def in_city(self, city: str) -> "PersonAddressBuilder":
        """
        Add city
        """
        self.person.city = city
        return self


if __name__ == "__main__":
    pb = PersonBuilder()
    test_person = (
        pb.lives.at("123 London Road")
        .in_city("London")
        .with_postcode("SW12BC")
        .works.at("Fabrikam")
        .as_a("Engineer")
        .earning(123000)
        .build()
    )
    print(test_person)

File: Go/Builder/builder_2.go
package main

import "fmt"

type Person struct {
	StreetAddress, Postcode, City string
	CompanyName, Position         string
	AnnualIncome                  int
}

type PersonBuilder struct {
	person *Person
}

func (b *PersonBuilder) Lives() *PersonAddressBuilder {
	return &PersonAddressBuilder{*b}
}

func (b *PersonBuilder) Works() *PersonJobBuilder {
	return &PersonJobBuilder{*b}
}

type PersonAddressBuilder struct {
	PersonBuilder
}

type PersonJobBuilder struct {
	PersonBuilder
}

func (pab *PersonAddressBuilder) At(streetAddress string) *PersonAddressBuilder {
	pab.person.StreetAddress = streetAddress
	return pab
}

func (pab *PersonAddressBuilder) In(city string) *PersonAddressBuilder {
	pab.person.City = city
	return pab
}

func (pab *PersonAddressBuilder) Postcode(postcode string) *PersonAddressBuilder {
	pab.person.Postcode = postcode
	return pab
}

func (pjb *PersonJobBuilder) At(companyName string) *PersonJobBuilder {
	pjb.person.CompanyName = companyName
	return pjb
}

func (pjb *PersonJobBuilder) AsA(position string) *PersonJobBuilder {
	pjb.person.Position = position
	return pjb
}

func (pjb *PersonJobBuilder) Earning(annualIncome int) *PersonJobBuilder {
	pjb.person.AnnualIncome = annualIncome
	return pjb
}

func NewPersonBuilder() *PersonBuilder {
	return &PersonBuilder{&Person{}}
}

func (b *PersonBuilder) Build() *Person {
	return b.person
}

func main() {
	pb := NewPersonBuilder()
	pb.Lives().At("123 London Road").In("London").Postcode("SW123BC").Works().At("Fabrikam").AsA("Programmer").Earning(123000)
	person := pb.Build()
	fmt.Println(person)
}
